### PR TITLE
Spring data jpa

### PR DESCRIPTION
**Pull request by AGuliaiev**

### CODE REVIEW

File: src/main/java/org/example/springintro/SpringIntroApplication.java [hunk 1]
```diff
@@ -0,0 +1,12 @@
+package org.example.springintro;
+
+import org.springframework.boot.SpringApplication;
+import org.springframework.boot.autoconfigure.SpringBootApplication;
+
+@SpringBootApplication
+public class SpringIntroApplication {
+
```
Comment (Elena-Bruyako) [inline src/main/java/org/example/springintro/SpringIntroApplication.java]:
```suggestion
```

File: pom.xml [hunk 2]
```diff
@@ -9,13 +9,15 @@
         <relativePath/> <!-- lookup parent from repository -->
     </parent>
     <groupId>org.example</groupId>
-    <artifactId>Spring_intro</artifactId>
+    <artifactId>spring-intro</artifactId>
```
Comment (Elena-Bruyako) [inline pom.xml:12]: didn't find required dependency
Look at HW description: Налаштувати залежності. Переконайся, що pom.xml містить залежність Liquibase Core і Liquibase Maven Plugin:

File: src/main/resources/application.properties [hunk 3]
```diff
@@ -3,5 +3,5 @@ spring.datasource.username=root
 spring.datasource.password=root
 spring.datasource.driver-class-name=com.mysql.cj.jdbc.Driver
 
-spring.jpa.hibernate.ddl-auto=create-drop
+spring.jpa.hibernate.ddl-auto=validate
```
Comment (Elena-Bruyako) [inline src/main/resources/application.properties:6]: add `server.servlet.context-path=/api`

File: src/main/java/org/example/springintro/model/Book.java [hunk 4]
```diff
@@ -1,29 +1,30 @@
...
 import jakarta.persistence.Id;
+import jakarta.persistence.Table;
 import lombok.Data;
+import org.hibernate.annotations.SQLDelete;
+import org.hibernate.annotations.Where;
 
 @Entity
 @Data
+@SQLDelete(sql = "UPDATE books SET is_delete = true WHERE id=?")
+@Where(clause = "is_delete=false")
```
Comment (Elena-Bruyako) [inline src/main/java/org/example/springintro/model/Book.java]: `@Where` is deprecated, use `@SQLRestriction`

File: src/main/java/org/example/springintro/model/Book.java [hunk 5]
```diff
@@ -1,29 +1,30 @@
...
 import jakarta.persistence.GeneratedValue;
 import jakarta.persistence.GenerationType;
 import jakarta.persistence.Id;
+import jakarta.persistence.Table;
 import lombok.Data;
+import org.hibernate.annotations.SQLDelete;
+import org.hibernate.annotations.Where;
 
 @Entity
 @Data
```
Comment (Elena-Bruyako) [inline src/main/java/org/example/springintro/model/Book.java]: Don't use `@Data` for entities

File: src/main/java/org/example/springintro/model/Book.java [hunk 6]
```diff
@@ -1,29 +1,30 @@
...
-//title (String, not null)
-//author (String, not null)
-//isbn (String, not null, unique)
-//price (BigDecimal, not null)
-//description (String)
-//coverImage (String)
     private Long id;
     private String title;
     private String author;
     private String isbn;
```
Comment (Elena-Bruyako) [inline src/main/java/org/example/springintro/model/Book.java:30]: Add constraints: nullable = false, unique - true
Check other fields according to the task description

File: src/main/java/org/example/springintro/model/Book.java [hunk 7]
```diff
@@ -1,29 +1,30 @@
...
     private String price;
     private String description;
     private String coverImage;
-
+    @Column(nullable = false)
```
Comment (Elena-Bruyako) [inline src/main/java/org/example/springintro/model/Book.java:35]:
```suggestion
```

File: src/main/java/org/example/springintro/repository/BookRepository.java [hunk 8]
```diff
@@ -0,0 +1,18 @@
+package org.example.springintro.repository;
+
+import java.util.List;
+import java.util.Optional;
+import org.example.springintro.model.Book;
+import org.springframework.data.jpa.repository.JpaRepository;
+import org.springframework.data.jpa.repository.Query;
+
+public interface BookRepository extends JpaRepository<Book, Long> {
+    Book save(Book book);
```
Comment (Elena-Bruyako) [inline src/main/java/org/example/springintro/repository/BookRepository.java]: All these methods provide JpaRepository

File: src/main/java/org/example/springintro/services/impl/BookServiceImpl.java [hunk 9]
```diff
@@ -0,0 +1,63 @@
...
+import org.example.springintro.model.Book;
+import org.example.springintro.repository.BookRepository;
+import org.example.springintro.services.BookService;
+import org.springframework.beans.factory.annotation.Autowired;
+import org.springframework.stereotype.Service;
+
+@RequiredArgsConstructor
+@Service
+public class BookServiceImpl implements BookService {
+    @Autowired
```
Comment (Elena-Bruyako) [inline src/main/java/org/example/springintro/services/impl/BookServiceImpl.java]: Redundant

File: src/main/java/org/example/springintro/services/impl/BookServiceImpl.java [hunk 10]
```diff
@@ -0,0 +1,63 @@
...
+        return bookMapper.toDto(book);
+    }
+
+    @Override
+    public BookDto updateById(Long id, CreateBookRequestDto requestDto) {
+        Book book = bookRepository.findById(id)
+                .orElseThrow(
+                        () -> new EntityNotFoundException("Book not found by id: " + id)
+                );
+        book.setTitle(requestDto.getTitle());
```
Comment (Elena-Bruyako) [inline src/main/java/org/example/springintro/services/impl/BookServiceImpl.java]: You can use Mapper method 
`bookMapper.updateBookFromDto(requestDto, book);`

where 
`void updateBookFromDto(CreateBookRequestDto book, @MappingTarget Book entity);`

File: src/main/resources/db/changelog/changes/01-create-books-table.yaml [hunk 11]
```diff
@@ -0,0 +1,33 @@
...
+            tableName: books
+            columns:
+              - column:
+                  name: id
+                  type: bigint
+                  autoIncrement: true
+                  constraints:
+                    primaryKey: true
+                    nullable: false
+              - column:
```
Comment (Elena-Bruyako) [inline src/main/resources/db/changelog/changes/01-create-books-table.yaml:16]: Add required constraints for columns according to the task description

File: src/main/resources/db/changelog/changes/02-add-delete-column-to-book-table.yaml [hunk 12]
```diff
@@ -0,0 +1,14 @@
+databaseChangeLog:
+  - changeSet:
+      id: add-delete-column-to-book-table
+      author: gulini
+      changes:
+        - addColumn:
+            tableName: books
+            columns:
+              - column:
```
Comment (Elena-Bruyako) [inline src/main/resources/db/changelog/changes/02-add-delete-column-to-book-table.yaml]: This column should be in 01-create-books-table.yaml

**Review by Elena-Bruyako — CHANGES_REQUESTED** [review]
Good job! Let's improve your solution!

File: pom.xml [hunk 13]
```diff
@@ -54,11 +56,26 @@
             <artifactId>mysql-connector-java</artifactId>
             <version>8.0.33</version>
         </dependency>
+        <dependency>
+            <groupId>org.liquibase</groupId>
+            <artifactId>liquibase-core</artifactId>
+            <version>${liquibase.version}</version>
```
Comment (Elena-Bruyako) [inline pom.xml]: spring boot parent already provides liquibase version

File: src/main/java/org/example/springintro/model/Book.java (hunk 6 above)
Comment (Elena-Bruyako) [reply src/main/java/org/example/springintro/model/Book.java:30]: Not fixed

Comment (Elena-Bruyako) [reply src/main/java/org/example/springintro/model/Book.java:30]: Screenshot 2024-05-01 at 08 19 12

File: src/main/resources/db/changelog/changes/01-create-books-table.yaml (hunk 11 above)
Comment (Elena-Bruyako) [reply src/main/resources/db/changelog/changes/01-create-books-table.yaml:16]: Not fixed

**Review by Elena-Bruyako — CHANGES_REQUESTED** [review]

File: src/main/java/org/example/springintro/model/Book.java [hunk 14]
```diff
@@ -0,0 +1,37 @@
...
+import lombok.Getter;
+import lombok.Setter;
+import org.hibernate.annotations.SQLDelete;
+import org.hibernate.annotations.SQLRestriction;
+
+@Entity
+@Setter
+@Getter
+@SQLDelete(sql = "UPDATE books SET is_delete = true WHERE id=?")
+@SQLRestriction(value = "is_delete=false")
```
Comment (okuzan) [inline src/main/java/org/example/springintro/model/Book.java:19]:
```suggestion
@SQLDelete(sql = "UPDATE books SET is_delete = TRUE WHERE id=?")
@SQLRestriction(value = "is_delete = FALSE")
```

File: src/main/java/org/example/springintro/services/impl/BookServiceImpl.java [hunk 15]
```diff
@@ -0,0 +1,56 @@
...
+        Book book = bookMapper.toModel(requestDto);
+        return bookMapper.toDto(bookRepository.save(book));
+    }
+
+    @Override
+    public List<BookDto> findAll() {
+        return bookRepository.findAll().stream()
+                .map(bookMapper::toDto)
+                .toList();
+    }
```
Comment (okuzan) [inline src/main/java/org/example/springintro/services/impl/BookServiceImpl.java:31]: you can use mapstruct to map list to list automatically, without streams

**Review by okuzan — APPROVED** [review]

**Pull request merged**